File: runme/mcgen/local/datan/datsrc/lsqmar.f
      SUBROUTINE LSQMAR(USERFN,T,Y,DELTAY,N,NR,NRED,LIST,X,CX,R,A,
     +SCRAT,NSTEP)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER(NMAX=1000)
      DIMENSION A(N,NRED),T(N),Y(N),DELTAY(N),X(NR)
      DIMENSION SCRAT(NRED,NRED),CX(NRED,NRED)
      DIMENSION C(NMAX),X1(NMAX),X2(NMAX),X1RED(NMAX),X2RED(NMAX)
      DIMENSION LIST(NR)
      COMMON /DASV02/ C,X1,X2,X1RED,X2RED
      PARAMETER(MAXSTP=100,EPSILN=1.D-8,TT=1.D-15,ALAMS=1.D-3)
      LOGICAL NEW,OK,FINAL,COVMAT
      EXTERNAL USERFN
      COVMAT=.TRUE.
      IF(NSTEP.LT.0) THEN
        COVMAT=.FALSE.
        NSTEP=ABS(NSTEP)
      END IF
      IF(NSTEP.LT.1) NSTEP=MAXSTP
      ALAM=ALAMS
C Initial value of minimum function
      R=DBLE(0.)
      DO 10 I=1,N
        R=R+((Y(I)-USERFN(X,NR,T(I)))/DELTAY(I))**2
   10 CONTINUE
      IF(NRED.LE.0) THEN
        NSTEP=1
        GO TO 100
      END IF
C For NR=NRED :  set LIST
      IF(NR.EQ.NRED) THEN
        DO 20 I=1,NR
          LIST(I)=1
   20   CONTINUE
      END IF
C start iteration
      FINAL=.FALSE.
      DO 90 ISTEP=1,NSTEP
C Numerical Derivatives
        CALL AUXDRI(USERFN,X,T,N,NR,NRED,LIST,A,OK)
        IF(.NOT.OK) THEN
          NSTEP=-3
          GO TO 100
        END IF
        DO 40 I=1,N
          DO 30 K=1,NRED
            A(I,K)=A(I,K)/DELTAY(I)
   30     CONTINUE
   40   CONTINUE
        DO 50 I=1,N
          C(I)=(Y(I)-USERFN(X,NR,T(I)))/DELTAY(I)
   50   CONTINUE
        IF(FINAL) THEN
C Final Step
          CALL MTXSVD(A,C,X1RED,R,N,NRED,1,DBLE(0.),OK)
          IF(.NOT.OK) THEN
            NSTEP=-1
            GO TO 100
          END IF
          CALL MTXZRV(X1,NR)
          CALL MTXPSV(X1,X1RED,NR,NRED,LIST)
          CALL MTXADV(X,X1,X,NR)
C Compute covariance matrix CX of unknowns
          IF(COVMAT) THEN
            CALL AUXDRI(USERFN,X,T,N,NR,NRED,LIST,A,OK)
            IF(.NOT.OK) THEN
              NSTEP=-3
              GO TO 100
            END IF
            DO 70 I=1,N
              DO 60 K=1,NRED
                A(I,K)=A(I,K)/DELTAY(I)
   60         CONTINUE
   70       CONTINUE
            CALL MTXMAT(A,A,CX,NRED,N,NRED)
            CALL MTXCHI(CX,SCRAT,NRED)
          END IF
          NSTEP=ISTEP
          GO TO 100
        END IF
C compute minimum function for two values of lambda (ALAM)
        CALL MTXMAR(A,C,ALAM,X1RED,X2RED,N,NRED,DBLE(0.),OK)
        IF(.NOT.OK) THEN
          NSTEP=-1
          GO TO 100
        END IF
        CALL MTXZRV(X1,NR)
        CALL MTXZRV(X2,NR)
        CALL MTXPSV(X1,X1RED,NR,NRED,LIST)
        CALL MTXPSV(X2,X2RED,NR,NRED,LIST)
        CALL MTXADV(X1,X,X1,NR)
        CALL MTXADV(X2,X,X2,NR)
        R1=DBLE(0.)
        R2=DBLE(0.)
        DO 80 I=1,N
          R1=R1+((Y(I)-USERFN(X1,NR,T(I)))/DELTAY(I))**2
          R2=R2+((Y(I)-USERFN(X2,NR,T(I)))/DELTAY(I))**2
   80   CONTINUE
        IF(.NOT.OK) THEN
          NSTEP=-1
          GO TO 100
        END IF
C evaluate results
        IF(R2.LE.R+TT) THEN
C reduce lambda and accept new point
          ALAM=ALAM*DBLE(0.1)
          CALL MTXCPV(X2,X,NR)
          RMIN=R2
          NEW=.TRUE.
        ELSE IF (R2.GT.R+TT .AND. R1.LE.R+TT) THEN
C keep current value of lambda and accept new point
          CALL MTXCPV(X1,X,NR)
          RMIN=R1
          NEW=.TRUE.
        ELSE
C increase lambda and reject new point
          ALAM=ALAM*DBLE(10.)
          NEW=.FALSE.
        END IF
        IF(NEW) THEN
C test for break-off criterion
          IF(ABS(R-RMIN) .LT. EPSILN*ABS(RMIN)+TT) THEN
            FINAL=.TRUE.
          ELSE
            R=RMIN
          END IF
        END IF
   90 CONTINUE
      NSTEP=-2
  100 RETURN
      END

File: runme/mcgen/local/datan/datsrc/lsqmar.c
#include <stdlib.h>
#include <limits.h>
#include "datsrc.h"

/* Common Block Declarations */

/*
struct {
    double c[1000], x1[1000], x2[1000], x1red[1000], x2red[1000];
} dasv02_;

#define dasv02_1 dasv02_
*/

void LIBFUNC lsqmar_(double (CBFUNC P_T *userfn)(double const P_T *,
             integer const P_T *, double const P_T *), double const P_T *t,
             double const P_T *y, double const P_T *deltay,
             integer const P_T *n, integer const P_T *nr, integer P_T *nred,
             integer P_T * list, double P_T *x, double P_T *cx, double P_T *r,
             double P_T *a, double P_T *scrat, integer P_T *nstep)
{
    /* Table of constant values */

    integer c__1 = 1;
    double c_d0 = 0.;

    /* System generated locals */
    integer a_dim1, a_offset, scrat_dim1, scrat_offset, cx_dim1, cx_offset,
            i__1, i__2, i__3;
    double d__1;

    /* Local variables */
    integer i, k, istep;
    double alam, rmin, r1, r2;
    logical final, new__, ok, covmat;
    double *c, *x1, *x2, *x1red, *x2red;
    unsigned long amem;

    if ((amem = sizeof(double) * *n) > MAXALLOC) {
        MEMERROR("lsqmar_: c (too large)");
        goto f_0;
    }
    c = (double *)MEMALLOC(amem);
    if (!c) {
        MEMERROR("lsqmar_: c");
        goto f_0;
    }
    if ((amem = sizeof(double) * 2 * (*nr + *nred)) > MAXALLOC) {
        MEMERROR("lsqmar_: x1 (too large)");
        goto f_1;
    }
    x1 = (double *)MEMALLOC(amem);
    if (!x1) {
        MEMERROR("lsqmar_: x1");
        goto f_1;
    }
    x2 = x1 + *nr; x1red = x2 + *nr; x2red = x1red + *nred;

    /* Parameter adjustments */
    --deltay;
    --y;
    --t;
    --x;
    --list;
    scrat_dim1 = *nred;
    scrat_offset = scrat_dim1 + 1;
    scrat -= scrat_offset;
    a_dim1 = *n;
    a_offset = a_dim1 + 1;
    a -= a_offset;
    cx_dim1 = *nred;
    cx_offset = cx_dim1 + 1;
    cx -= cx_offset;

    /* Function Body */
    covmat = TRUE_;
    if (*nstep < 0) {
        covmat = FALSE_;
        *nstep = abs(*nstep);
    }
    if (*nstep < 1) {
        *nstep = 100;
    }
    alam = .001;
/* Initial value of minimum function */
    *r = 0.;
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
/* Computing 2nd power */
        d__1 = (y[i] - (*userfn)(&x[1], nr, &t[i])) / deltay[i];
        *r += d__1 * d__1;
    }
    if (*nred <= 0) {
        *nstep = 1;
        goto L100;
    }
/* For NR=NRED :  set LIST */
    if (*nr == *nred) {
        i__1 = *nr;
        for (i = 1; i <= i__1; ++i) {
            list[i] = 1;
        }
    }
/* start iteration */
    final = FALSE_;
    i__1 = *nstep;
    for (istep = 1; istep <= i__1; ++istep) {
/* Numerical Derivatives */
        auxdri_(userfn, &x[1], &t[1], n, nr, nred, &list[1], &a[a_offset], &ok);
        if (! ok) {
            *nstep = -3;
            goto L100;
        }
        i__2 = *n;
        for (i = 1; i <= i__2; ++i) {
            i__3 = *nred;
            for (k = 1; k <= i__3; ++k) {
                a[i + k * a_dim1] /= deltay[i];
            }
        }
        i__2 = *n;
        for (i = 1; i <= i__2; ++i) {
            c[i - 1] = (y[i] - (*userfn)(&x[1], nr, &t[i])) / deltay[i];
        }
        if (final) {
/* Final Step */
            mtxsvd_(&a[a_offset], c, x1red, r, n, nred, &c__1, &c_d0, &ok);
            if (! ok) {
                *nstep = -1;
                goto L100;
            }
            mtxzrv_(x1, nr);
            mtxpsv_(x1, x1red, nr, nred, &list[1]);
            mtxadv_(&x[1], x1, &x[1], nr);
/* Compute covariance matrix CX of unknowns */
            if (covmat) {
                auxdri_(userfn, &x[1], &t[1], n, nr, nred, &list[1],
                       &a[a_offset], &ok);
                if (! ok) {
                    *nstep = -3;
                    goto L100;
                }
                i__2 = *n;
                for (i = 1; i <= i__2; ++i) {
                    i__3 = *nred;
                    for (k = 1; k <= i__3; ++k) {
                        a[i + k * a_dim1] /= deltay[i];
                    }
                }
                mtxmat_(&a[a_offset], &a[a_offset], &cx[cx_offset], nred, n,
                        nred);
                mtxchi_(&cx[cx_offset], &scrat[scrat_offset], nred);
            }
            *nstep = istep;
            goto L100;
        }
/* compute minimum function for two values of lambda (ALAM) */
        mtxmar_(&a[a_offset], c, &alam, x1red,
                x2red, n, nred, &c_d0, &ok);
        if (! ok) {
            *nstep = -1;
            goto L100;
        }
        mtxzrv_(x1, nr);
        mtxzrv_(x2, nr);
        mtxpsv_(x1, x1red, nr, nred, &list[1]);
        mtxpsv_(x2, x2red, nr, nred, &list[1]);
        mtxadv_(x1, &x[1], x1, nr);
        mtxadv_(x2, &x[1], x2, nr);
        r1 = 0.;
        r2 = 0.;
        i__2 = *n;
        for (i = 1; i <= i__2; ++i) {
/* Computing 2nd power */
            d__1 = (y[i] - (*userfn)(x1, nr, &t[i])) / deltay[i];
            r1 += d__1 * d__1;
/* Computing 2nd power */
            d__1 = (y[i] - (*userfn)(x2, nr, &t[i])) / deltay[i];
            r2 += d__1 * d__1;
        }
        if (! ok) {
            *nstep = -1;
            goto L100;
        }
/* evaluate results */
        if (r2 <= *r + 1e-15) {
/* reduce lambda and accept new point */
            alam *= .1;
            mtxcpv_(x2, &x[1], nr);
            rmin = r2;
            new__ = TRUE_;
        } else if (r2 > *r + 1e-15 && r1 <= *r + 1e-15) {
/* keep current value of lambda and accept new point */
            mtxcpv_(x1, &x[1], nr);
            rmin = r1;
            new__ = TRUE_;
        } else {
/* increase lambda and reject new point */
            alam *= 10.;
            new__ = FALSE_;
        }
        if (new__) {
/* test for break-off criterion */
            if ((d__1 = *r - rmin, abs(d__1)) < abs(rmin) * 1e-8 + 1e-15) {
                final = TRUE_;
            } else {
                *r = rmin;
            }
        }
    }
    *nstep = -2;
L100:
    MEMFREE((void *)x1);
f_1:
    MEMFREE((void *)c);
f_0:
    return;
} /* lsqmar_ */

